### PR TITLE
Solidus 3 preparation

### PR DESCRIPTION
**Pull request by cpfergus1**
This prepares the extension to work with current supported versions of Solidus.
This PR satisfies #51 

Discussion points:
Use of conditionals and if parameters to satisfy versioning requirements.
Other possible parameters to use for SoftDelete checks than Solidus Version

### CODE REVIEW

File: app/controllers/spree/admin/sale_prices_controller.rb [hunk 1]
```diff
@@ -22,7 +22,7 @@ def create
 
       def destroy
         @sale_price = Spree::SalePrice.find(params[:id])
-        @sale_price.discard
+        @sale_price.send(SolidusSalePrices::SoftDelete.delete_method)
```
Comment (jarednorman) [inline app/controllers/spree/admin/sale_prices_controller.rb:25]: Should probably use `public_send` throughout, just as a matter of safe practices, even if it has no impact to the overall function here.

File: app/models/spree/sale_price.rb [hunk 2]
```diff
@@ -1,9 +1,18 @@
 module Spree
   class SalePrice < ActiveRecord::Base
-    include Spree::SoftDeletable
+    if self.respond_to?(:acts_as_paranoid)
```
Comment (jarednorman) [inline app/models/spree/sale_price.rb]: Won't this be true within any application that has paranoia loaded, regardless of the soft deletion method available in Solidus?

**Review by jarednorman — CHANGES_REQUESTED** [review]

File: app/helpers/solidus_sale_prices/soft_delete.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal :true
+
+module SolidusSalePrices
+  module SoftDelete
+    def self.delete_method
```
Comment (kennyadsl) [inline app/helpers/solidus_sale_prices/soft_delete.rb:5]: I'm not sure we need this. What if we just always use Discard methods? If I recall correctly, in Solidus 2.10 we had support for both Discard and Paranoia just for backward compatibility, so using discard methods would work exactly the same without emitting deprecation warnings. Did you already try?

**Review by kennyadsl — COMMENTED** [review]
Maybe there's the opportunity of removing the conditional code for soft-deletion, I left a comment about it.

File: app/helpers/solidus_sale_prices/soft_delete.rb (hunk 3 above)
Comment (cpfergus1) [reply app/helpers/solidus_sale_prices/soft_delete.rb:5]: In Solidus 2.10, Spree::SoftDeletable has not been added yet. This prevents the use of discard methods which is the reason for the conditional methods and I thought it might be cleaner to but some of that logic in a helper

Comment (jarednorman) [reply app/helpers/solidus_sale_prices/soft_delete.rb:5]: Even without `Spree::SoftDeletable`, discard is still present. I don't see any need to use paranoia.

Comment (cpfergus1) [reply app/helpers/solidus_sale_prices/soft_delete.rb:5]: I will explore this further, I was under the impression that the functionality of discard was tied to Spree::SoftDeletable

**Review by cpfergus1 — COMMENTED** [review]
Just started playing with it this morning and found that the specs do fail with just the discard methods implemented. It works and is arguably more concise than the previous implementation if I just force the SoftDeleteable functionality in the price model:
```
    if ::Spree.solidus_gem_version >= Gem::Version.new('2.11')
      include Spree::SoftDeletable
    else
      require 'discard'
      acts_as_paranoid
      include Discard::Model
      self.discard_column = :deleted_at
    end
```
Further, I could throw that into the helper module and place a note to remove when 2.10 is no longer supported as well as create an issue in this repository.

Additionally, without paranoia, the following fails and requires further investigation:
```
  1) Spree::Variant with a valid sale destroys all sale prices when it is destroyed
     Failure/Error:
       expect { variant.discard }
         .to change { Spree::SalePrice.all.size }
         .from(3).to(0)
     
       expected `Spree::SalePrice.all.size` to have changed from 3 to 0, but did not change
     # ./spec/models/variant_spec.rb:144:in `block (3 levels) in <top (required)>'

  2) Spree::Price#destroy when there are sale prices associated to the price destroys all sale prices when it is destroyed
     Failure/Error:
       expect { price.discard }
         .to change { Spree::SalePrice.all.size }
         .from(1).to(0)
     
       expected `Spree::SalePrice.all.size` to have changed from 1 to 0, but did not change
     # ./spec/models/price_spec.rb:92:in `block (4 levels) in <top (required)>'
```
Without Discard::Model, none of the discard methods work.

I will continue to look for other solutions and please let me know if I am steering off in the wrong direction.
Thanks!

File: app/controllers/spree/admin/sale_prices_controller.rb (hunk 1 above)
Comment (cpfergus1) [reply app/controllers/spree/admin/sale_prices_controller.rb]: Removed send in favor of #discard. Please let me know if the new implementation works!

File: app/models/spree/sale_price.rb (hunk 2 above)
Comment (cpfergus1) [reply app/models/spree/sale_price.rb]: Has been removed in favor of Solidus version to be more explicit. Please let me know if you have any other ideas in which to based this conditional off of.

File: app/helpers/solidus_sale_prices/soft_delete.rb (hunk 3 above)
Comment (cpfergus1) [reply app/helpers/solidus_sale_prices/soft_delete.rb]: Just started playing with it this morning and found that the specs do fail with just the discard methods implemented. It works and is arguably more concise than the previous implementation if I just force the SoftDeleteable functionality in the price model:
```
    if ::Spree.solidus_gem_version >= Gem::Version.new('2.11')
      include Spree::SoftDeletable
    else
      require 'discard'
      acts_as_paranoid
      include Discard::Model
      self.discard_column = :deleted_at
    end
```
Further, I could throw that into the helper module and place a note to remove when 2.10 is no longer supported as well as create an issue in this repository.

Additionally, without paranoia, the following fails and requires further investigation:
```
  1) Spree::Variant with a valid sale destroys all sale prices when it is destroyed
     Failure/Error:
       expect { variant.discard }
         .to change { Spree::SalePrice.all.size }
         .from(3).to(0)
     
       expected `Spree::SalePrice.all.size` to have changed from 3 to 0, but did not change
     # ./spec/models/variant_spec.rb:144:in `block (3 levels) in <top (required)>'

  2) Spree::Price#destroy when there are sale prices associated to the price destroys all sale prices when it is destroyed
     Failure/Error:
       expect { price.discard }
         .to change { Spree::SalePrice.all.size }
         .from(1).to(0)
     
       expected `Spree::SalePrice.all.size` to have changed from 1 to 0, but did not change
     # ./spec/models/price_spec.rb:92:in `block (4 levels) in <top (required)>'
```
Without Discard::Model, none of the discard methods work.

I will continue to look for other solutions and please let me know if I am steering off in the wrong direction.
Thanks!

**Review by kennyadsl — APPROVED** [review]

**Comment by kennyadsl** [comment]
@jarednorman time for a second pass?

File: .byebug_history [hunk 4]
```diff
@@ -0,0 +1,3 @@
+exit
+::Spree.solidus_gem_version
+::Spree.solidus_gem_version > Gem::Version.new('2.10')
```
Comment (jarednorman) [inline .byebug_history]: This shouldn't be committed.

File: Gemfile [hunk 5]
```diff
@@ -6,11 +6,6 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 branch = ENV.fetch('SOLIDUS_BRANCH', 'master')
 gem 'solidus', github: 'solidusio/solidus', branch: branch
 
-# Needed to help Bundler figure out how to resolve dependencies,
-# otherwise it takes forever to resolve them.
-# See https://github.com/bundler/bundler/issues/6677
-gem 'rails', '>0.a'
-
```
Comment (jarednorman) [inline Gemfile:13]: Why is this removed?

File: app/models/spree/sale_price.rb [hunk 6]
```diff
@@ -1,6 +1,11 @@
 module Spree
   class SalePrice < ActiveRecord::Base
-    include Spree::SoftDeletable
+
+    if ::Spree.solidus_gem_version >= Gem::Version.new('2.11')
+      include Spree::SoftDeletable
+    else
+      include SolidusSalePrices::SoftDelete
```
Comment (jarednorman) [inline app/models/spree/sale_price.rb:7]: I don't see the point in having this in a concern and including it in only one place. This code is temporary until all supported versions of Solidus have `Spree::SoftDeletable`, so let's make this code easier to delete and just inline it.

File: lib/solidus_sale_prices/engine.rb [hunk 7]
```diff
@@ -4,7 +4,7 @@
 
 module SolidusSalePrices
   class Engine < Rails::Engine
-    include SolidusSupport::EngineExtensions::Decorators
+    include SolidusSupport::EngineExtensions
```
Comment (jarednorman) [inline lib/solidus_sale_prices/engine.rb:7]: Typo in commit message: "depreciated" should be "deprecated".

File: spec/features/admin/sale_prices_spec.rb [hunk 8]
```diff
@@ -78,14 +78,12 @@
...
       fill_in('sale_price_value', with: 32.33)
-
       click_button('Create')
-
       within all('[data-hook="sale_prices"] [data-hook="prices_row"]').last do
         expect(page).to have_content(variant.descriptive_name)
         expect(page).to have_content(32.33)
         expect(page).to have_content(country.name)
-        expect(page).to have_content(variant.prices.last.currency)
+        expect(page).to have_content(variant.prices.first.currency)
```
Comment (jarednorman) [inline spec/features/admin/sale_prices_spec.rb:86]: Unless I'm missing something, this order isn't guaranteed. Either way, the tests should know the currency, so why don't we just make the assertion against the literal?

```suggestion
        expect(page).to have_content("EUR")
```

File: app/helpers/solidus_sale_prices/soft_delete.rb [hunk 9]
```diff
@@ -9,5 +9,6 @@ module SoftDelete
       acts_as_paranoid
       include Discard::Model
       self.discard_column = :deleted_at
+    end
```
Comment (jarednorman) [inline app/helpers/solidus_sale_prices/soft_delete.rb:12]: I'd appreciate a pass to clean up the commit sequence, so changes like this are in the right commits.

File: spec/models/variant_spec.rb [hunk 10]
```diff
@@ -24,18 +22,15 @@
   end
 
   it 'changes the price for each specific currency' do
-    variant = create(:multi_price_variant)
-
     variant.prices.each do |p|
-      variant.put_on_sale 10.95, { currencies: [ p.currency ] }
+     variant.put_on_sale 10.95, { currencies: [ p.currency ] }
```
Comment (jarednorman) [inline spec/models/variant_spec.rb:26]: This block should be indented two spaces, not one.

File: spec/models/sale_price_spec.rb [hunk 11]
```diff
@@ -39,7 +39,7 @@
     end
 
     it 'sets the end time' do
-      expect(sale_price.end_at).to be_within(1.second).of(Time.now)
+      expect(sale_price.end_at).to be_within(2.second).of(Time.now)
```
Comment (jarednorman) [inline spec/models/sale_price_spec.rb:42]: I don't actually care if you fix, but fwiw in the commit message "accomidate" should be "accommodate".

**Review by jarednorman — CHANGES_REQUESTED** [review]
Overall approach seems good to me. This could use some rebase love to get rid of some churn and things committed in the wrong commits to keep the history sensible.

File: .byebug_history (hunk 4 above)
Comment (cpfergus1) [reply .byebug_history]: Absolutely, Thanks for pointing that out

File: Gemfile (hunk 5 above)
Comment (cpfergus1) [reply Gemfile:13]: No longer required. There used to be an issue with bundler versions below 2.2 I believe which caused hanging when resolving dependencies for older versions of rails which this line of code rectified. The bug in bundler has been rectified so I was just removing it since I was working on the repo.

File: lib/solidus_sale_prices/engine.rb (hunk 7 above)
Comment (cpfergus1) [reply lib/solidus_sale_prices/engine.rb:7]: Thank you for pointing that out!

File: spec/models/sale_price_spec.rb (hunk 11 above)
Comment (cpfergus1) [reply spec/models/sale_price_spec.rb:42]: Thank you for catching that, I will get it rectified.

File: app/models/spree/sale_price.rb (hunk 6 above)
Comment (cpfergus1) [reply app/models/spree/sale_price.rb:7]: I see your point on this and looking back at it now, I reimplemented the same functionality of Spree::SoftDeleteable so should we even include the conditional at all? I am thinking of just leaving
```
include SolidusSalePrices::SoftDelete
```
and open an issue to replace that with Spree::SoftDeleteable once 2.10 is no longer supported. What do you think @jarednorman?

Comment (jarednorman) [reply app/models/spree/sale_price.rb:7]: I think there's something subtle that I missed on my first review: `SolidusSalePrices::SoftDelete` uses paranoia, but `Spree::SoftDeletable` does not. This means that using the former means you can call `.destroy` on the record and cause it to get soft-deleted, whereas with the latter doing so will cause the record to actually be dropped from the database.

If we care about this difference, I'd say we maintain the current functionality and leave a note in the codebase regarding how we make the breaking change in the future to move to the soft-deletion functionality that comes with Solidus:

```ruby
# The following code enables soft-deletion. In Solidus v2.11+ there is a mixin
# that handles soft-deletion consistently across all Solidus records that need
# it. Once we no longer support v2.10, we can remove the dependency on paranoia
# and replace these lines with:
#
#     include Spree::SoftDeletable
#
# However, this will be a breaking change as it will change the behaviour of
# calling `destroy` on these records.
acts_as_paranoid
include Discard::Model
self.discard_column = :deleted_at
```

File: app/helpers/solidus_sale_prices/soft_delete.rb (hunk 9 above)
Comment (cpfergus1) [reply app/helpers/solidus_sale_prices/soft_delete.rb:12]: Done!

**Comment by cpfergus1** [comment]
Thank you for your corrections @kennyadsl and @jarednorman. All current topics have been remedied and pushed. I also removed two older commits that were obsolete and cleaned up the history a little bit to be more explicit. Please let me know if I left anything out!
